content/algorithms/pagerank/pagerank.md

```diff
@@ -0,0 +1,513 @@
+
++++
+
+## Stationary Distributions
+
++++
+
+In a Markov chain, a stationary distribution represents the long-term behavior of the system. It is a probability distribution that remains unchanged after transitions within the Markov chain. The stationary distribution is reached when the Markov chain reaches a steady state, where the probabilities of transitioning between different web pages no longer change significantly. 
+
+In the context of the PageRank algorithm, the stationary distribution corresponds to the importance or rank assigned to each web page. This distribution provides a measure of the relative importance of each web page, as it represents the probability of a random surfer visiting each page in the long run.
+
+Considering the Markov chain we plotted above, let's compute the probability distribution across states over time. The initial probability distribution across states is taken as uniform. At each step we perform the operation $\pi_{n+1} = \pi_{n}.P$
+
+```{code-cell}
+probability_distribution = np.full(4, 0.25)
+for i in range(25):
+    print("Step", i, ": ", probability_distribution)
+    probability_distribution = np.dot(probability_distribution, P)
+```
+
+It is observed that towards the end of iterations, the probability distribution is reaching a steady state i.e. the stationary distribution.
+
+This brute-force method of computing stationary distributions by performing the operation $\pi_{n+1} = \pi_{n}.P$ at every step is called the Power Iteration method. We continue the iterations till the difference between $\pi_{n+1}$ and $\pi_{n}$ becomes less than the tolerance threshold value.
+
+There exist algebraic methods that can compute stationary distributions in a more direct manner. But practical experiments on large web networks have found that the brute-force method has the best performance  because most web networks have relatively fast convergence times (~50 iterations to reach a reasonable tolerance threshold).
+
++++
+
+## Existence of Stationary Distributions
+
++++
+
+<i>Do all Markov chains converge to a Stationary distribution?</i>
+
+To answer this question, let us look at two examples.
+
+```{code-cell}
+I = nx.MultiDiGraph()
+I.add_edges_from([(1, 1), (0, 0)])
+
+plt.figure(figsize=(4, 2))
+
+pos = {0: (1, 1), 1: (2, 1)}
+nx.draw(
+    I,
+    pos,
+    with_labels=True,
+    node_size=1000,
+    node_color="yellow",
+    connectionstyle="arc3, rad = 0.1",
+)
+el = nx.draw_networkx_edge_labels(I, pos, edge_labels={(1, 1): 1.0, (0, 0): 1.0})
+plt.title("G1", y=-0.2)
+plt.show()
+```
+
+Both states in above Markov chain have only one incoming link that too from itself. Thus, it contains isolated subsets of states that cannot be reached from the rest of the chain. This is called a <b><i>reducible</i></b> Markov chain 
+
+Here the transition probability matrix is an identity matrix because of which every initial distribution becomes a stationary distribution.
```

Comment:
```suggestion
Here the transition probability matrix is an identity matrix because every initial distribution is a stationary distribution. The state are sometimes called "absorbing states" because the system never leaves those states.
```